docs/hello_nf-core/03_add_module.md

```diff
@@ -1,63 +1,893 @@
+
+```groovy
+input:
+tuple val(meta), path(files_in)
+
+output:
+tuple val(meta), path("${prefix}"), emit: file_out
+```
+
+This pattern is standard across all nf-core modules. The metadata map (commonly called `meta`) is a Groovy-style map containing information about a sample or dataset, with `id` being the required field used for naming outputs and tracking samples.
+
+Why use metadata maps?
+
+- **Sample tracking**: Keep sample information with data throughout the workflow
+- **Standardization**: All nf-core modules follow this pattern
+- **Flexibility**: Easy to add custom metadata fields
+- **Output naming**: Consistent file naming based on sample IDs
+
+!!! note "Learn more about metadata"
+
+    For a comprehensive introduction to working with metadata in Nextflow workflows, including how to read metadata from samplesheets and use it to customize processing, see the [Metadata in workflows](../side_quests/metadata) side quest.
+
+For now, we'll pass the output from `CAT_CAT` to `cowpy` with the character parameter. In the next section, we'll adapt `cowpy` to follow nf-core conventions.
+
+### 1.9. Wire up CAT_CAT in the workflow
+
+Now we need to modify our workflow code to use `CAT_CAT` instead of `collectGreetings`. Since `CAT_CAT` requires metadata tuples, we need to create a metadata map and combine it with our files.
+
+Open [core-hello/workflows/hello.nf](core-hello/workflows/hello.nf) and modify the workflow logic in the `main` block:
+
+=== "After"
+
+    ```groovy title="core-hello/workflows/hello.nf" linenums="26" hl_lines="7-17"
+        // emit a greeting
+        sayHello(ch_samplesheet)
+
+        // convert the greeting to uppercase
+        convertToUpper(sayHello.out)
+
+        // collect all the greetings into one file using nf-core cat/cat module
+        // create metadata map with batch name as the ID
+        def cat_meta = [ id: params.batch ]
+        ch_for_cat = convertToUpper.out.collect().map { files -> tuple(cat_meta, files) }
+
+        CAT_CAT(ch_for_cat)
+
+        // generate ASCII art of the greetings with cowpy
+        // Extract the file from the tuple since cowpy doesn't use metadata yet
+        ch_for_cowpy = CAT_CAT.out.file_out.map{ meta, file -> file }
+        cowpy(ch_for_cowpy, params.character)
+    ```
+
+=== "Before"
+
+    ```groovy title="core-hello/workflows/hello.nf" linenums="26" hl_lines="7-15"
+        // emit a greeting
+        sayHello(ch_samplesheet)
+
+        // convert the greeting to uppercase
+        convertToUpper(sayHello.out)
+
+        // collect all the greetings into one file
+        collectGreetings(convertToUpper.out.collect(), params.batch)
+
+        // emit a message about the size of the batch
+        collectGreetings.out.count.view { "There were $it greetings in this batch" }
+
+        // generate ASCII art of the greetings with cowpy
+        cowpy(collectGreetings.out.outfile, params.character)
+    ```
+
+Let's break down what we changed:
```

Comment:
I wonder if we should break this down into separate steps, because there's a lot going on here. In particular I got confused about how we added the metamap because I missed it among all of that.